provider/oci/instance.go

```diff
@@ -4,44 +4,297 @@
 package oci
 
 import (
+	"context"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
 	"github.com/juju/errors"
+	"github.com/juju/juju/status"
 
-	"github.com/juju/juju/environs/context"
+	envcontext "github.com/juju/juju/environs/context"
+	"github.com/juju/juju/environs/instances"
 	"github.com/juju/juju/instance"
 	"github.com/juju/juju/network"
+
+	ociCore "github.com/oracle/oci-go-sdk/core"
 )
 
-type ociInstance struct{}
+const (
+	// MinVolumeSizeMB is the minimum size in MB for a volume or boot disk
+	MinVolumeSizeMB = 51200
+
+	// MaxVolumeSizeMB is the maximum size in MB for a volume or boot disk
+	MaxVolumeSizeMB = 16777216
+)
+
+type ociInstance struct {
+	arch     string
+	instType *instances.InstanceType
+	env      *Environ
+	mutex    sync.Mutex
+	etag     *string
+	raw      ociCore.Instance
+}
+
+type vnicWithIndex struct {
+	Vnic ociCore.Vnic
+	Idx  int
+}
 
 var _ instance.Instance = (*ociInstance)(nil)
-var _ instance.InstanceFirewaller = (*ociInstance)(nil)
+
+var statusMap map[ociCore.InstanceLifecycleStateEnum]status.Status = map[ociCore.InstanceLifecycleStateEnum]status.Status{
+	ociCore.InstanceLifecycleStateProvisioning:  status.Provisioning,
+	ociCore.InstanceLifecycleStateRunning:       status.Running,
+	ociCore.InstanceLifecycleStateStarting:      status.Provisioning,
+	ociCore.InstanceLifecycleStateStopping:      status.Running,
+	ociCore.InstanceLifecycleStateStopped:       status.Running,
+	ociCore.InstanceLifecycleStateCreatingImage: status.Provisioning,
+	ociCore.InstanceLifecycleStateTerminating:   status.Running,
+	ociCore.InstanceLifecycleStateTerminated:    status.Running,
+}
+
+// newInstance returns a new oracleInstance
+func newInstance(raw ociCore.Instance, env *Environ) (*ociInstance, error) {
+	if raw.Id == nil {
+		return nil, errors.New(
+			"Instance response does not contain an ID",
+		)
+	}
+	instance := &ociInstance{
+		raw:  raw,
+		env:  env,
+		arch: "amd64",
+	}
+
+	return instance, nil
+}
+
+// SetInstance sets the raw property of ociInstance{}
+// Testing purposes.
+func (o *ociInstance) SetInstance(inst ociCore.Instance) {
+	o.raw = inst
+}
+
+func (o *ociInstance) availabilityZone() string {
+	return *o.raw.AvailabilityDomain
+}
 
 // Id implements instance.Instance
 func (o *ociInstance) Id() instance.Id {
-	return instance.Id(0)
+	return instance.Id(*o.raw.Id)
 }
 
 // Status implements instance.Instance
-func (o *ociInstance) Status(ctx context.ProviderCallContext) instance.InstanceStatus {
-	return instance.InstanceStatus{}
+func (o *ociInstance) Status(ctx envcontext.ProviderCallContext) instance.InstanceStatus {
+	// This should not happen, unless someone bypassed newInstance()
+	// and created the ociInstance{} object manually.
+	if err := o.refresh(); err != nil {
+		return instance.InstanceStatus{}
+	}
+	state, ok := statusMap[o.raw.LifecycleState]
+	if !ok {
+		state = status.Empty
```

Comment:
I think we have a status.Unknown which we could alternately use.

Reply:
That's what I initially used, but it seems that Unknown is meant for applications and units.Looking at it again, I see ```KnownInstanceStatus()``` references ```Pending``` and ```Unknown``` as well. Changing to ```Unknown```